File: electron/vault/trash.ts
import fs from 'node:fs';
import path from 'node:path';
import { resolveVaultPath } from './index';
import { parseFrontmatter } from './markdown';
import { titleFromFilename } from './sanitize';
import type { NoteData, NoteSummary } from './notes';

const TRASH_DIR = '.trash';
const TRASH_META_FILE = '.trash-meta.json';

export interface TrashMeta {
  items: Array<{
    id: string;
    originalPath: string;
    trashedAt: string;
  }>;
}

function readTrashMeta(vaultRoot: string): TrashMeta {
  const metaPath = path.join(vaultRoot, TRASH_DIR, TRASH_META_FILE);
  try {
    const raw = fs.readFileSync(metaPath, 'utf-8');
    return JSON.parse(raw) as TrashMeta;
  } catch {
    return { items: [] };
  }
}

function writeTrashMeta(vaultRoot: string, meta: TrashMeta): void {
  const trashDir = path.join(vaultRoot, TRASH_DIR);
  fs.mkdirSync(trashDir, { recursive: true });
  const metaPath = path.join(trashDir, TRASH_META_FILE);
  fs.writeFileSync(metaPath, JSON.stringify(meta, null, 2), 'utf-8');
}

/**
 * Moves a .md file to .trash/, updates .trash-meta.json with original path and timestamp.
 */
export function softDelete(vaultPath: string, notePath: string): boolean {
  const resolved = resolveVaultPath(vaultPath);
  const srcFullPath = path.join(resolved, notePath);

  if (!fs.existsSync(srcFullPath)) return false;

  // Read the note to get its id
  const fileContent = fs.readFileSync(srcFullPath, 'utf-8');
  const { data } = parseFrontmatter(fileContent);

  // Ensure .trash/ exists
  const trashDir = path.join(resolved, TRASH_DIR);
  fs.mkdirSync(trashDir, { recursive: true });

  // Move file to .trash/
  const filename = path.basename(notePath);
  const destPath = path.join(trashDir, filename);

  // Handle filename conflicts in trash
  let finalFilename = filename;
  if (fs.existsSync(destPath)) {
    const ext = path.extname(filename);
    const base = path.basename(filename, ext);
    let counter = 2;
    while (fs.existsSync(path.join(trashDir, `${base} ${counter}${ext}`))) {
      counter++;
    }
    finalFilename = `${base} ${counter}${ext}`;
  }

  fs.renameSync(srcFullPath, path.join(trashDir, finalFilename));

  // Update trash meta
  const meta = readTrashMeta(resolved);
  meta.items.push({
    id: data.id,
    originalPath: notePath,
    trashedAt: new Date().toISOString(),
  });
  writeTrashMeta(resolved, meta);

  return true;
}

/**
 * Reads original path from meta, moves file back to its original location.
 * If targetNotebook is provided, restores to that notebook instead.
 */
export function restore(
  vaultPath: string,
  trashFilename: string,
  targetNotebook?: string
): NoteData {
  const resolved = resolveVaultPath(vaultPath);
  const trashDir = path.join(resolved, TRASH_DIR);
  const trashFilePath = path.join(trashDir, trashFilename);

  if (!fs.existsSync(trashFilePath)) {
    throw new Error(`Trashed file not found: ${trashFilename}`);
  }

  // Read the file content
  const fileContent = fs.readFileSync(trashFilePath, 'utf-8');
  const { data, content } = parseFrontmatter(fileContent);

  // Find the meta entry
  const meta = readTrashMeta(resolved);
  const metaIndex = meta.items.findIndex(
    (item) => item.id === data.id || path.basename(item.originalPath) === trashFilename
  );

  let restorePath: string;

  if (targetNotebook) {
    // Restore to specified notebook
    const targetDir = path.join(resolved, targetNotebook);
    fs.mkdirSync(targetDir, { recursive: true });
    restorePath = path.join(targetNotebook, trashFilename);
  } else if (metaIndex >= 0) {
    // Restore to original path
    restorePath = meta.items[metaIndex].originalPath;
    const restoreDir = path.join(resolved, path.dirname(restorePath));
    if (!fs.existsSync(restoreDir)) {
      throw new Error(
        `Original notebook directory no longer exists: ${path.dirname(restorePath)}`
      );
    }
  } else {
    throw new Error(`No metadata found for trashed file: ${trashFilename}`);
  }

  const fullRestorePath = path.join(resolved, restorePath);

  // Ensure parent directory exists
  fs.mkdirSync(path.dirname(fullRestorePath), { recursive: true });

  // Move file back
  fs.renameSync(trashFilePath, fullRestorePath);

  // Remove meta entry
  if (metaIndex >= 0) {
    meta.items.splice(metaIndex, 1);
    writeTrashMeta(resolved, meta);
  }

  const title = titleFromFilename(path.basename(restorePath));
  const notebook = path.dirname(restorePath).split(path.sep)[0];

  return {
    id: data.id,
    title,
    content,
    path: restorePath,
    notebook,
    tags: data.tags,
    created: data.created,
    modified: data.modified,
    isTrashed: false,
  };
}

/**
 * Permanently removes a file from .trash/ and its meta entry.
 */
export function permanentDelete(vaultPath: string, trashFilename: string): boolean {
  const resolved = resolveVaultPath(vaultPath);
  const trashDir = path.join(resolved, TRASH_DIR);
  const filePath = path.join(trashDir, trashFilename);

  if (!fs.existsSync(filePath)) return false;

  // Read file to get id for meta removal
  const fileContent = fs.readFileSync(filePath, 'utf-8');
  const { data } = parseFrontmatter(fileContent);

  // Delete the file
  fs.unlinkSync(filePath);

  // Remove meta entry
  const meta = readTrashMeta(resolved);
  meta.items = meta.items.filter((item) => item.id !== data.id);
  writeTrashMeta(resolved, meta);

  return true;
}

/**
 * Deletes all files in .trash/, clears meta, returns count of deleted files.
 */
export function emptyTrash(vaultPath: string): number {
  const resolved = resolveVaultPath(vaultPath);
  const trashDir = path.join(resolved, TRASH_DIR);

  if (!fs.existsSync(trashDir)) return 0;

  const entries = fs.readdirSync(trashDir);
  let count = 0;

  for (const entry of entries) {
    if (entry === TRASH_META_FILE) continue;
    const filePath = path.join(trashDir, entry);
    try {
      fs.unlinkSync(filePath);
      count++;
    } catch {
      // Skip files that can't be deleted
    }
  }

  // Clear meta
  writeTrashMeta(resolved, { items: [] });

  return count;
}

/**
 * Lists trashed notes with original path info.
 */
export function listTrash(vaultPath: string): NoteSummary[] {
  const resolved = resolveVaultPath(vaultPath);
  const trashDir = path.join(resolved, TRASH_DIR);

  if (!fs.existsSync(trashDir)) return [];

  const meta = readTrashMeta(resolved);
  const results: NoteSummary[] = [];

  const entries = fs.readdirSync(trashDir);
  for (const entry of entries) {
    if (entry === TRASH_META_FILE) continue;
    if (!entry.endsWith('.md')) continue;

    const filePath = path.join(trashDir, entry);
    try {
      const fileContent = fs.readFileSync(filePath, 'utf-8');
      const { data, content } = parseFrontmatter(fileContent);
      const title = titleFromFilename(entry);

      // Find original path from meta
      const metaItem = meta.items.find((item) => item.id === data.id);
      const originalNotebook = metaItem
        ? path.dirname(metaItem.originalPath).split(path.sep)[0]
        : '.trash';

      results.push({
        id: data.id,
        title,
        path: path.join('.trash', entry),
        notebook: originalNotebook,
        tags: data.tags,
        created: data.created,
        modified: data.modified,
        snippet: content.slice(0, 200).trim(),
      });
    } catch {
      // Skip files that can't be parsed
    }
  }

  results.sort((a, b) => b.modified.localeCompare(a.modified));
  return results;
}

File: src-tauri/src/vault/trash.rs
use crate::vault::{
    markdown::parse_frontmatter,
    notes::NoteData,
    resolve_vault_path,
    sanitize::resolve_filename_conflict,
    TRASH_DIR, TRASH_META_FILE,
};
use serde::{Deserialize, Serialize};
use std::fs;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrashMeta {
    pub items: Vec<TrashItem>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TrashItem {
    pub id: String,
    pub original_path: String,
    pub trashed_at: String,
}

fn read_trash_meta(vault_root: &std::path::Path) -> TrashMeta {
    let meta_path = vault_root.join(TRASH_DIR).join(TRASH_META_FILE);
    fs::read_to_string(&meta_path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or(TrashMeta { items: vec![] })
}

fn write_trash_meta(vault_root: &std::path::Path, meta: &TrashMeta) -> Result<(), String> {
    let trash_dir = vault_root.join(TRASH_DIR);
    fs::create_dir_all(&trash_dir).map_err(|e| e.to_string())?;
    let meta_path = trash_dir.join(TRASH_META_FILE);
    let json = serde_json::to_string_pretty(meta).map_err(|e| e.to_string())?;
    fs::write(&meta_path, json).map_err(|e| e.to_string())
}

/// Moves a `.md` file to `.trash/`, updates `.trash-meta.json`.
pub fn soft_delete(vault_path: &str, note_rel_path: &str) -> Result<bool, String> {
    let resolved = resolve_vault_path(vault_path);
    let src = resolved.join(note_rel_path);

    if !src.exists() {
        return Err(format!("Note not found: {}", note_rel_path));
    }

    let trash_dir = resolved.join(TRASH_DIR);
    fs::create_dir_all(&trash_dir).map_err(|e| e.to_string())?;

    let filename = src
        .file_name()
        .and_then(|n| n.to_str())
        .ok_or("Invalid filename")?
        .to_string();
    let dest_filename = resolve_filename_conflict(&trash_dir, &filename);
    let dest = trash_dir.join(&dest_filename);

    // Read note id before moving
    let content = fs::read_to_string(&src).unwrap_or_default();
    let parsed = parse_frontmatter(&content);
    let note_id = parsed.data.id;

    fs::rename(&src, &dest).map_err(|e| e.to_string())?;

    // Update trash meta
    let mut meta = read_trash_meta(&resolved);
    meta.items.push(TrashItem {
        id: note_id,
        original_path: note_rel_path.to_string(),
        trashed_at: chrono::Utc::now().to_rfc3339(),
    });
    write_trash_meta(&resolved, &meta)?;

    Ok(true)
}

/// Soft-deletes an entire notebook directory to `.trash/`.
pub fn soft_delete_dir(vault_path: &str, notebook_rel_path: &str) -> Result<bool, String> {
    let resolved = resolve_vault_path(vault_path);
    let src = resolved.join(notebook_rel_path);

    if !src.exists() {
        return Err(format!("Notebook not found: {}", notebook_rel_path));
    }

    let trash_dir = resolved.join(TRASH_DIR);
    fs::create_dir_all(&trash_dir).map_err(|e| e.to_string())?;

    let name = src
        .file_name()
        .and_then(|n| n.to_str())
        .ok_or("Invalid notebook path")?
        .to_string();
    let dest = trash_dir.join(&name);
    fs::rename(&src, &dest).map_err(|e| e.to_string())?;

    Ok(true)
}

/// Restores a trashed note to `target_notebook` (or its original notebook).
pub fn restore_note(
    vault_path: &str,
    trash_path: &str,
    target_notebook: Option<&str>,
) -> Result<NoteData, String> {
    let resolved = resolve_vault_path(vault_path);
    let src = resolved.join(TRASH_DIR).join(
        std::path::Path::new(trash_path)
            .file_name()
            .ok_or("Invalid trash path")?,
    );

    if !src.exists() {
        return Err(format!("Trashed note not found: {}", trash_path));
    }

    // Determine restore destination
    let mut meta = read_trash_meta(&resolved);
    let filename = src
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("")
        .to_string();

    let notebook = if let Some(nb) = target_notebook {
        nb.to_string()
    } else {
        // Look up original path in trash meta
        meta.items
            .iter()
            .find(|item| {
                std::path::Path::new(&item.original_path)
                    .file_name()
                    .and_then(|n| n.to_str())
                    == Some(&filename)
            })
            .and_then(|item| {
                std::path::Path::new(&item.original_path)
                    .parent()
                    .map(|p| p.to_string_lossy().to_string())
            })
            .unwrap_or_else(|| "Inbox".to_string())
    };

    let dest_dir = resolved.join(&notebook);
    fs::create_dir_all(&dest_dir).map_err(|e| e.to_string())?;

    let dest_filename = resolve_filename_conflict(&dest_dir, &filename);
    let dest = dest_dir.join(&dest_filename);
    fs::rename(&src, &dest).map_err(|e| e.to_string())?;

    // Remove from trash meta
    meta.items.retain(|item| {
        std::path::Path::new(&item.original_path)
            .file_name()
            .and_then(|n| n.to_str())
            != Some(&filename)
    });
    write_trash_meta(&resolved, &meta)?;

    let rel_path = dest
        .strip_prefix(&resolved)
        .map_err(|e| e.to_string())?
        .to_string_lossy()
        .to_string();

    crate::vault::notes::get_note(vault_path, &rel_path)
}

/// Permanently deletes a note from the trash.
pub fn permanent_delete(vault_path: &str, trash_path: &str) -> Result<bool, String> {
    let resolved = resolve_vault_path(vault_path);
    let filename = std::path::Path::new(trash_path)
        .file_name()
        .and_then(|n| n.to_str())
        .ok_or("Invalid trash path")?
        .to_string();
    let file = resolved.join(TRASH_DIR).join(&filename);

    if file.exists() {
        fs::remove_file(&file).map_err(|e| e.to_string())?;
    }

    // Remove from meta
    let mut meta = read_trash_meta(&resolved);
    meta.items.retain(|item| {
        std::path::Path::new(&item.original_path)
            .file_name()
            .and_then(|n| n.to_str())
            != Some(&filename)
    });
    write_trash_meta(&resolved, &meta)?;

    Ok(true)
}

/// Permanently deletes all notes in the trash. Returns the count of deleted notes.
pub fn empty_trash(vault_path: &str) -> Result<usize, String> {
    let resolved = resolve_vault_path(vault_path);
    let trash_dir = resolved.join(TRASH_DIR);

    if !trash_dir.exists() {
        return Ok(0);
    }

    let mut count = 0usize;
    for entry in fs::read_dir(&trash_dir).map_err(|e| e.to_string())? {
        let entry = match entry {
            Ok(e) => e,
            Err(_) => continue,
        };
        let path = entry.path();
        if path.extension().and_then(|e| e.to_str()) == Some("md") {
            if fs::remove_file(&path).is_ok() {
                count += 1;
            }
        }
    }

    // Clear meta
    write_trash_meta(&resolved, &TrashMeta { items: vec![] })?;
    Ok(count)
}
